File: backend/pkg/healthkit/metrics_test.go
package healthkit

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestLookup_KnownQuantityType(t *testing.T) {
	t.Parallel()

	metric, ok := Lookup("HKQuantityTypeIdentifierHeartRate")
	require.True(t, ok)
	require.Equal(t, "heart_rate", metric.MetricType)
	require.Equal(t, KindQuantity, metric.Kind)
	require.Equal(t, "count/min", metric.CanonicalUnit)
}

func TestLookup_KnownCategoryType(t *testing.T) {
	t.Parallel()

	metric, ok := Lookup("HKCategoryTypeIdentifierSleepAnalysis")
	require.True(t, ok)
	require.Equal(t, "sleep_stage", metric.MetricType)
	require.Equal(t, KindCategory, metric.Kind)
	require.Empty(t, metric.CanonicalUnit)
}

// An identifier the server does not know must report false rather than error, so the caller can store
// the sample verbatim instead of discarding the user's data.
func TestLookup_UnknownTypeIsNotAnError(t *testing.T) {
	t.Parallel()

	metric, ok := Lookup("HKQuantityTypeIdentifierSomethingAppleShippedLastTuesday")
	require.False(t, ok)
	require.Empty(t, metric.MetricType)
}

func TestLookupByMetricType(t *testing.T) {
	t.Parallel()

	metric, ok := LookupByMetricType("heart_rate")
	require.True(t, ok)
	require.Equal(t, KindQuantity, metric.Kind)

	_, ok = LookupByMetricType("not_a_metric")
	require.False(t, ok)

	_, ok = LookupByMetricType("  ")
	require.False(t, ok)
}

func TestLookup_TrimsWhitespace(t *testing.T) {
	t.Parallel()

	metric, ok := Lookup("  HKQuantityTypeIdentifierStepCount\n")
	require.True(t, ok)
	require.Equal(t, "step_count", metric.MetricType)
}

func TestNormalizeUnit(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		hkType   string
		unit     string
		expected string
		ok       bool
	}{
		{name: "healthkit spelling", hkType: "HKQuantityTypeIdentifierHeartRate", unit: "count/min", expected: "count/min", ok: true},
		{name: "alias collapses to canonical", hkType: "HKQuantityTypeIdentifierHeartRate", unit: "bpm", expected: "count/min", ok: true},
		{name: "case insensitive", hkType: "HKQuantityTypeIdentifierBloodPressureSystolic", unit: "MMHG", expected: "mmHg", ok: true},
		{name: "surrounding whitespace", hkType: "HKQuantityTypeIdentifierBodyMass", unit: " kg ", expected: "kg", ok: true},
		{name: "wrong unit rejected", hkType: "HKQuantityTypeIdentifierBodyMass", unit: "lb", ok: false},
		{name: "unit of another metric rejected", hkType: "HKQuantityTypeIdentifierHeartRate", unit: "mmHg", ok: false},
		{name: "empty unit rejected", hkType: "HKQuantityTypeIdentifierHeartRate", unit: "", ok: false},
	}

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			metric, found := Lookup(tt.hkType)
			require.True(t, found)

			normalized, ok := metric.NormalizeUnit(tt.unit)
			require.Equal(t, tt.ok, ok)
			require.Equal(t, tt.expected, normalized)
		})
	}
}

func TestNormalizeCategoryValue(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		value    string
		expected string
		ok       bool
	}{
		{name: "canonical name", value: "asleepCore", expected: "asleepCore", ok: true},
		{name: "case insensitive", value: "ASLEEPREM", expected: "asleepREM", ok: true},
		{name: "raw healthkit enum", value: "4", expected: "asleepDeep", ok: true},
		{name: "in bed enum zero", value: "0", expected: "inBed", ok: true},
		{name: "legacy asleep alias", value: "asleep", expected: "asleepUnspecified", ok: true},
		{name: "full enum constant", value: "HKCategoryValueSleepAnalysisAwake", expected: "awake", ok: true},
		{name: "unknown stage rejected", value: "dreaming", ok: false},
		{name: "out of range enum rejected", value: "99", ok: false},
		{name: "empty rejected", value: "", ok: false},
	}

	metric, found := Lookup("HKCategoryTypeIdentifierSleepAnalysis")
	require.True(t, found)

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			normalized, ok := metric.NormalizeCategoryValue(tt.value)
			require.Equal(t, tt.ok, ok)
			require.Equal(t, tt.expected, normalized)
		})
	}
}

// Every quantity metric must accept its own canonical unit, and every category metric must accept its
// own canonical values. A typo in the table above would otherwise reject valid samples at runtime.
func TestMetricTable_IsSelfConsistent(t *testing.T) {
	t.Parallel()

	for hkType, metric := range metrics {
		require.NotEmpty(t, metric.MetricType, "%s has no metric type", hkType)

		switch metric.Kind {
		case KindQuantity:
			require.NotEmpty(t, metric.CanonicalUnit, "%s has no canonical unit", hkType)
			require.Empty(t, metric.allowedValues, "%s is a quantity metric but lists category values", hkType)

			normalized, ok := metric.NormalizeUnit(metric.CanonicalUnit)
			require.True(t, ok, "%s rejects its own canonical unit", hkType)
			require.Equal(t, metric.CanonicalUnit, normalized)
		case KindCategory:
			require.Empty(t, metric.CanonicalUnit, "%s is a category metric but declares a unit", hkType)
			require.NotEmpty(t, metric.allowedValues, "%s has no category values", hkType)
			require.Empty(t, metric.acceptedUnits, "%s is a category metric but accepts units", hkType)

			for _, allowed := range metric.allowedValues {
				normalized, ok := metric.NormalizeCategoryValue(allowed.canonical)
				require.True(t, ok, "%s rejects its own value %s", hkType, allowed.canonical)
				require.Equal(t, allowed.canonical, normalized)
			}
		default:
			t.Fatalf("%s has unknown kind %q", hkType, metric.Kind)
		}
	}
}

File: src/app/healthkit/__tests__/metrics.test.ts
import { describe, expect, it } from 'vitest';
import { allMetrics, lookup, lookupByMetricType, normalizeCategoryValue, normalizeUnit } from '../metrics.js';

describe('lookup', () => {
  it('maps a known quantity type', () => {
    const metric = lookup('HKQuantityTypeIdentifierHeartRate');
    expect(metric?.metricType).toBe('heart_rate');
    expect(metric?.kind).toBe('quantity');
    expect(metric?.canonicalUnit).toBe('count/min');
  });

  it('maps a known category type', () => {
    const metric = lookup('HKCategoryTypeIdentifierSleepAnalysis');
    expect(metric?.metricType).toBe('sleep_stage');
    expect(metric?.kind).toBe('category');
    expect(metric?.canonicalUnit).toBe('');
  });

  it('reports unknown types as missing rather than throwing, so the caller can store them verbatim', () => {
    expect(lookup('HKQuantityTypeIdentifierSomethingAppleShippedLastTuesday')).toBeUndefined();
  });

  it('trims whitespace', () => {
    expect(lookup('  HKQuantityTypeIdentifierStepCount\n')?.metricType).toBe('step_count');
  });
});

describe('lookupByMetricType', () => {
  it('finds a known name and refuses an empty or unknown one', () => {
    expect(lookupByMetricType('heart_rate')?.kind).toBe('quantity');
    expect(lookupByMetricType('not_a_metric')).toBeUndefined();
    expect(lookupByMetricType('  ')).toBeUndefined();
  });
});

describe('normalizeUnit', () => {
  const heart = lookup('HKQuantityTypeIdentifierHeartRate')!;
  const mass = lookup('HKQuantityTypeIdentifierBodyMass')!;
  const bp = lookup('HKQuantityTypeIdentifierBloodPressureSystolic')!;

  it('accepts HealthKit spelling and aliases, collapsing them to the canonical unit', () => {
    expect(normalizeUnit(heart, 'count/min')).toBe('count/min');
    expect(normalizeUnit(heart, 'bpm')).toBe('count/min');
    expect(normalizeUnit(bp, 'MMHG')).toBe('mmHg');
    expect(normalizeUnit(mass, ' kg ')).toBe('kg');
  });

  it('rejects a wrong unit, a unit of another metric, and the empty string', () => {
    expect(normalizeUnit(mass, 'lb')).toBeUndefined();
    expect(normalizeUnit(heart, 'mmHg')).toBeUndefined();
    expect(normalizeUnit(heart, '')).toBeUndefined();
  });
});

describe('normalizeCategoryValue', () => {
  const sleep = lookup('HKCategoryTypeIdentifierSleepAnalysis')!;

  it('accepts canonical names, aliases, and HealthKit integer enums', () => {
    expect(normalizeCategoryValue(sleep, 'asleepCore')).toBe('asleepCore');
    expect(normalizeCategoryValue(sleep, 'ASLEEPREM')).toBe('asleepREM');
    expect(normalizeCategoryValue(sleep, '4')).toBe('asleepDeep');
    expect(normalizeCategoryValue(sleep, '0')).toBe('inBed');
    expect(normalizeCategoryValue(sleep, 'asleep')).toBe('asleepUnspecified');
    expect(normalizeCategoryValue(sleep, 'HKCategoryValueSleepAnalysisAwake')).toBe('awake');
  });

  it('rejects an unknown stage, an out-of-range enum, and the empty string', () => {
    expect(normalizeCategoryValue(sleep, 'dreaming')).toBeUndefined();
    expect(normalizeCategoryValue(sleep, '99')).toBeUndefined();
    expect(normalizeCategoryValue(sleep, '')).toBeUndefined();
  });
});

describe('the metric table is self-consistent', () => {
  it('every quantity accepts its own canonical unit and every category accepts its own values', () => {
    for (const [hkType, metric] of allMetrics()) {
      expect(metric.metricType, `${hkType} has no metric type`).not.toBe('');
      if (metric.kind === 'quantity') {
        expect(metric.canonicalUnit, `${hkType} has no canonical unit`).not.toBe('');
        expect(metric.allowedValues, `${hkType} is a quantity metric but lists category values`).toHaveLength(0);
        expect(normalizeUnit(metric, metric.canonicalUnit)).toBe(metric.canonicalUnit);
      } else {
        expect(metric.canonicalUnit, `${hkType} is a category metric but declares a unit`).toBe('');
        expect(metric.allowedValues.length, `${hkType} has no category values`).toBeGreaterThan(0);
        expect(metric.acceptedUnits, `${hkType} is a category metric but accepts units`).toHaveLength(0);
        for (const allowed of metric.allowedValues) {
          expect(normalizeCategoryValue(metric, allowed.canonical)).toBe(allowed.canonical);
        }
      }
    }
  });
});
